x/gamm/pool-models/balancer/amm.go

```diff
@@ -281,39 +283,100 @@ func (p *Pool) CalcJoinPoolShares(_ sdk.Context, tokensIn sdk.Coins, swapFee sdk
 	}
 
 	// Add all exact coins we can (no swap). ctx arg doesn't matter for Balancer.
-	numShares, remCoins, err := cfmm_common.MaximalExactRatioJoin(p, sdk.Context{}, tokensIn)
+	numShares, remainingLiquidity, err := cfmm_common.MaximalExactRatioJoin(p, sdk.Context{}, tokensIn)
 	if err != nil {
 		return sdk.ZeroInt(), sdk.NewCoins(), err
 	}
 
-	// update liquidity for accurate calcSingleAssetJoin calculation
-	newLiquidity = tokensIn.Sub(remCoins)
-	for _, coin := range newLiquidity {
-		poolAsset := poolAssetsByDenom[coin.Denom]
-		poolAsset.Token.Amount = poolAssetsByDenom[coin.Denom].Token.Amount.Add(coin.Amount)
-		poolAssetsByDenom[coin.Denom] = poolAsset
-	}
-
-	newTotalShares := totalShares.Add(numShares)
+	// This is new liquidity that to be added to the pool from exact ratio join.
+	// If there is remainingLiqudity, we will add it to newLiquidity by
+	// performing single asset join on each coin in remainingLiquidity.
+	newLiquidity = tokensIn.Sub(remainingLiquidity)
 
 	// If there are coins that couldn't be perfectly joined, do single asset joins
 	// for each of them.
-	if !remCoins.Empty() {
-		for _, coin := range remCoins {
-			newShares, err := p.calcSingleAssetJoin(coin, swapFee, poolAssetsByDenom[coin.Denom], newTotalShares)
-			if err != nil {
-				return sdk.ZeroInt(), sdk.NewCoins(), err
-			}
-
-			newLiquidity = newLiquidity.Add(coin)
-			newTotalShares = newTotalShares.Add(newShares)
-			numShares = numShares.Add(newShares)
+	if !remainingLiquidity.Empty() {
+		// update pool assets for accurate calcSingleAssetJoin calculation
+		if err := updateIntermediaryPoolAssets(newLiquidity, poolAssetsByDenom); err != nil {
+			return sdk.ZeroInt(), sdk.NewCoins(), err
 		}
+
+		// update total shares for accurate calcSingleAssetJoin calculation
+		newTotalShares := totalShares.Add(numShares)
+
+		newNumSharesFromRemaining, newLiquidityFromRemaining, err := p.calcJoinSingleAssetTokensIn(remainingLiquidity, newTotalShares, poolAssetsByDenom, swapFee)
+		if err != nil {
+			return sdk.ZeroInt(), sdk.NewCoins(), err
+		}
+		numShares.Add(newNumSharesFromRemaining)
+		newLiquidity = newLiquidity.Add(newLiquidityFromRemaining...)
```

Comment:
Unit tests for `CalcJoinPoolShares` validating that these updates happen should be added in  a separate PR

Reply:
https://github.com/osmosis-labs/osmosis/issues/1722